File: architecture-skills/design-patterns/examples/observer/observer.php
<?php
// Observer pattern in PHP.
//
// PHP ships SplSubject/SplObserver in the standard library, but their
// notify() carries no payload, so real code usually rolls its own with
// closures — as here. subscribe() returns an unsubscribe closure.

declare(strict_types=1);

final class StockTicker
{
    /** @var array<int, Closure(string, float): void> */
    private array $observers = [];
    private int $nextId = 0;

    /** @param Closure(string, float): void $observer */
    public function subscribe(Closure $observer): Closure
    {
        $id = $this->nextId++;
        $this->observers[$id] = $observer;
        return function () use ($id): void {
            unset($this->observers[$id]);
        };
    }

    public function updatePrice(string $symbol, float $price): void
    {
        foreach ($this->observers as $observer) {
            $observer($symbol, $price);
        }
    }
}

final class PriceDisplay
{
    public function __construct(private readonly string $name)
    {
    }

    public function onPrice(string $symbol, float $price): void
    {
        printf("%s: %s is now $%.2f\n", $this->name, $symbol, $price);
    }
}

function makeAlert(string $watched, float $threshold): Closure
{
    return function (string $symbol, float $price) use ($watched, $threshold): void {
        if ($symbol === $watched && $price > $threshold) {
            printf("ALERT: %s crossed $%.2f (now $%.2f)\n", $watched, $threshold, $price);
        }
    };
}

$ticker = new StockTicker();
$display = new PriceDisplay('Lobby display');
$unsubscribe = $ticker->subscribe($display->onPrice(...));
$ticker->subscribe(makeAlert('ACME', 100.0));

$ticker->updatePrice('ACME', 98.5);
$ticker->updatePrice('ACME', 101.25);

$unsubscribe();
$ticker->updatePrice('GLOBEX', 45.0); // only the alert remains

File: architecture-skills/design-patterns/examples/observer/observer.rs
// Observer pattern in Rust.
//
// The GoF form (subject holds &mut references to observers) fights the
// borrow checker. This example stores boxed closures, which the ticker
// owns outright — simple, but observers can't be shared elsewhere or
// unsubscribed by identity. The alternative is Rc<RefCell<dyn Observer>>
// (shared, mutable observers at the cost of runtime borrow checks);
// in practice most Rust code uses channels (std::sync::mpsc) instead.

type PriceObserver = Box<dyn Fn(&str, f64)>;

struct StockTicker {
    observers: Vec<PriceObserver>,
}

impl StockTicker {
    fn new() -> Self {
        Self { observers: Vec::new() }
    }

    fn subscribe(&mut self, observer: PriceObserver) {
        self.observers.push(observer);
    }

    fn update_price(&self, symbol: &str, price: f64) {
        for observer in &self.observers {
            observer(symbol, price);
        }
    }
}

struct PriceDisplay {
    name: String,
}

impl PriceDisplay {
    fn on_price(&self, symbol: &str, price: f64) {
        println!("{}: {} is now ${:.2}", self.name, symbol, price);
    }
}

fn make_alert(watched: &str, threshold: f64) -> PriceObserver {
    let watched = watched.to_string();
    Box::new(move |symbol, price| {
        if symbol == watched && price > threshold {
            println!("ALERT: {watched} crossed ${threshold:.2} (now ${price:.2})");
        }
    })
}

fn main() {
    let mut ticker = StockTicker::new();

    let display = PriceDisplay {
        name: "Lobby display".to_string(),
    };
    // The closure takes ownership of the display; the ticker now owns both.
    ticker.subscribe(Box::new(move |symbol, price| display.on_price(symbol, price)));
    ticker.subscribe(make_alert("ACME", 100.0));

    ticker.update_price("ACME", 98.5);
    ticker.update_price("ACME", 101.25);
    ticker.update_price("GLOBEX", 45.0);
}
